### PR TITLE
gh-119879: Utilize last character gap for two-way periodic needles

### PR DESCRIPTION
**Pull request by dg-pb**
All is laid out in the issue. It would be great if someone familiar with an algorithm looked at this.

* Issue: gh-119879

### CODE REVIEW

**Comment by nineteendo** [comment]
cc @sweeneyde

**Comment by dgrigonis** [comment]
Removed asserts:
```c
assert(gap >= i - cut + 1);
assert(i - cut + 1 > gap);
```
as they are implied by
```c
Py_ssize_t gap_jump_end = Py_MIN(len_needle, cut + gap);
```

**Comment by nineteendo** [comment]
Note that asserts don't generate code:

```c
#define assert(op) /* empty */
```

File: Objects/stringlib/fastsearch.h [hunk 1]
```diff
@@ -256,7 +256,7 @@ STRINGLIB(_factorize)(const STRINGLIB_CHAR *needle,
 
        The local period of the cut is the minimal length of a string w
        such that (left endswith w or w endswith left)
-       and (right startswith w or w startswith left).
+       and (right startswith w or w startswith right).
```
Comment (sweeneyde) [inline Objects/stringlib/fastsearch.h:259]: Nice catch, thanks.

File: Objects/stringlib/fastsearch.h [hunk 2]
```diff
@@ -405,11 +406,18 @@ STRINGLIB(_two_way)(const STRINGLIB_CHAR *haystack, Py_ssize_t len_haystack,
...
                    (needle[len_needle - 1] & TABLE_MASK));
+
             Py_ssize_t i = Py_MAX(cut, memory);
             for (; i < len_needle; i++) {
                 if (needle[i] != window[i]) {
-                    LOG("Right half does not match.\n");
-                    window_last += i - cut + 1;
+                    if (i < gap_jump_end) {
+                        LOG("Early right half mismatch: jump by gap.\n");
+                        window_last += gap;
```
Comment (sweeneyde) [inline Objects/stringlib/fastsearch.h:415]: I'd like to keep the two assertions here and below--it makes it clearer to me that we're just doing `window_last += Py_MAX(gap, i - cut + 1);`.

I think this refactoring makes sense, though I want to do some extra testing to make sure the one extra branch doesn't slow things down in the (common) case where the mismatch is at the first `i`-value.

**Review by sweeneyde — COMMENTED** [review]
I think this is a good idea, thanks for investigating this.

I'd change the title to something like "Utilize last character gap for two-way periodic needles". The Horspool/"bad character" skip is everything involving `table`, and this change doesn't look to involve that. Using `gap` is the first "good suffix"-type rule.

Please also add a NEWS entry (you can click the "bedevere/news --> details" check).

File: Objects/stringlib/fastsearch.h (hunk 2 above)
Comment (dg-pb) [reply Objects/stringlib/fastsearch.h:415]: Added them back, but modified to better match `if` clause. Hope it's ok.

File: Objects/stringlib/fastsearch.h [hunk 3]
```diff
@@ -405,11 +406,20 @@ STRINGLIB(_two_way)(const STRINGLIB_CHAR *haystack, Py_ssize_t len_haystack,
             window = window_last - len_needle + 1;
             assert((window[len_needle - 1] & TABLE_MASK) ==
                    (needle[len_needle - 1] & TABLE_MASK));
+
```
Comment (nineteendo) [inline Objects/stringlib/fastsearch.h]:
```suggestion
```

**Comment by dgrigonis** [comment]
Suspended to double check one last thing to make sure it all works ok. All seems to be correct.

File: Misc/NEWS.d/next/Core and Builtins/2024-06-02-06-12-35.gh-issue-119879.Jit951.rst [hunk 4]
```diff
@@ -0,0 +1 @@
+Utilization of last character gap (good suffix rule) for two-way periodic needles.
```
Comment (nineteendo) [inline Misc/NEWS.d/next/Core and Builtins/2024-06-02-06-12-35.gh-issue-119879.Jit951.rst]: It's not clear we're talking about fast search here.

**Comment by sweeneyde** [comment]
(It looks like there's still whitespace on line 409.)

Other than that, I'm happy to merge this. Although the cases it speeds up are probably rare in practice, there is no real increase in code complexity here: you're aligning the periodic algorithm with an existing trick used in the non-periodic algorithm, so LGTM.

Going forward, we should discuss all of your ongoing work on fastsearch.h under one issue, say https://github.com/python/cpython/issues/119702 for example. There's no need to open a new issue for each individual proposed change.

**Pull request merged**